tests/interfaces/test_tf.py

```diff
@@ -768,3 +768,379 @@ def test_model_nested_tensor(self):
+        assert spy.call_args[1]["wrt"] == {0, 1}
+
+    def test_chained_gradient_value(self, mocker, tol):
+        """Test that the returned gradient value for two chained qubit QNodes
+        is correct."""
+        spy = mocker.spy(qml.qnodes.JacobianQNode, "jacobian")
+        dev1 = qml.device("default.qubit", wires=3)
+
+        @qml.qnode(dev1, interface="tf")
+        def circuit1(a, b, c):
+            qml.RX(a, wires=0)
+            qml.RX(b, wires=1)
+            qml.RX(c, wires=2)
+            qml.CNOT(wires=[0, 1])
+            qml.CNOT(wires=[1, 2])
+            return qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliY(2))
+
+        dev2 = qml.device("default.qubit", wires=2)
+
+        @qml.qnode(dev2, interface="tf")
+        def circuit2(data, weights):
+            qml.RX(data[0], wires=0)
+            qml.RX(data[1], wires=1)
+            qml.CNOT(wires=[0, 1])
+            qml.RZ(weights[0], wires=0)
+            qml.RZ(weights[1], wires=1)
+            qml.CNOT(wires=[0, 1])
+            return qml.expval(qml.PauliX(0) @ qml.PauliY(1))
+
+        def cost(a, b, c, weights):
+            return circuit2(circuit1(a, b, c), weights)
+
+        # Set the first parameter of circuit1 as non-differentiable.
+        a = tf.constant(0.4)
+
+        # the remaining free parameters are all differentiable
+        b = tf.Variable(0.5)
+        c = tf.Variable(0.1)
+        weights = tf.Variable([0.2, 0.3])
+
+        with tf.GradientTape() as tape:
+            loss = cost(a, b, c, weights)
+
+        grad = tape.gradient(loss, [b, c, weights])
+
+        # Output should have shape [dcost/db, dcost/dc, dcost/dw],
+        # where b,c are scalars, and w is a vector of length 2.
+        assert len(grad) == 3
+        assert grad[0].shape == tuple() # scalar
+        assert grad[1].shape == tuple() # scalar
+        assert grad[2].shape == (2,)    # vector
+
+        cacbsc = tf.cos(a)*tf.cos(b)*tf.sin(c)
+
+        expected = [
+            # analytic expression for dcost/db
+            -tf.cos(a)*tf.sin(b)*tf.sin(c)*tf.cos(cacbsc)*tf.sin(weights[0])*tf.sin(tf.cos(a)),
+            # analytic expression for dcost/dc
+            tf.cos(a)*tf.cos(b)*tf.cos(c)*tf.cos(cacbsc)*tf.sin(weights[0])*tf.sin(tf.cos(a)),
+            # analytic expression for dcost/dw[0]
+            tf.sin(cacbsc)*tf.cos(weights[0])*tf.sin(tf.cos(a)),
+            # analytic expression for dcost/dw[1]
+            0
+        ]
+
+        # np.hstack 'flattens' the ragged gradient array allowing it
+        # to be compared with the expected result
+        assert np.allclose(np.hstack(grad), expected, atol=tol, rtol=0)
+
+        # Check that the parameter-shift rule was applied
+        # to all parameters in circuit2 (i.e., wrt=None)
+        assert spy.call_args_list[0][1]["wrt"] == None
+
+        # check that the parameter-shift rule was not applied
+        # to the first parameter of circuit1
+        assert spy.call_args_list[1][1]["wrt"] == {1, 2}
```

Comment:
Since these tests were copied from the two parallel PRs, a lot of the suggestions that I had in those ones carry over here (specifically writing the asserts out directly)

Reply:
Have copied over the explicit asserts 👍